// EditorComponent/src/react/EditorApp.tsx
import { ChangeSettingsAction } from './actions/ChangeSettingAction';
import { EditorSettingsService } from './service/EditorSettingsService';
import { IEditorStateRecord, EditorStateFactory } from './model/EditorState';
import * as React from "react";
import { EditorSettingsComponent, EditorSettingsComponentProps } from "./EditorSettingsComponent"
import { ListComponent, ListItem, ListComponentProps } from "./ListComponent"
import * as immutable from "immutable"

export class EditorApp extends React.Component<any, IEditorStateRecord> {
    private editorSettingsService = new EditorSettingsService();

    constructor() {
        super();
        this.state = this.getInitialState();

        this.runNewParserClick = this.runNewParserClick.bind(this);
        this.runOriginalParserClick = this.runOriginalParserClick.bind(this);
        this.toggleSettings = this.toggleSettings.bind(this);
    }

    private getInitialState(): IEditorStateRecord {
        var initialState = EditorStateFactory();
        var editorSettings = this.editorSettingsService.createAndLoadSettingsFromCookies();
        initialState = initialState.set('settings', editorSettings);
        return initialState;
    }

    runNewParserClick() {
        console.log('runNewParserClick')
    }

    runOriginalParserClick() {
        console.log('runOriginalParserClick')
    }

    toggleSettings(name: string) {
        var newSettingsValue = this.state.settings.items.get(name) != true;
        var action = new ChangeSettingsAction(name, newSettingsValue);
        this.setState((x, y) => action.run(x));
    }

    render() {
        var style = { height: "300px", border: "1px solid grey" };

        return (
            <div className="row">
                <div className="col s4">
                    <div id="queryEditorContainer" style={style}></div>
                    <button className="btn align-right" onClick={this.runNewParserClick}>Run</button>
                    <button className="btn align-right" onClick={this.runOriginalParserClick}>Run Original</button>
                    <EditorSettingsComponent editorSettings={this.state.settings} toggleSettings={this.toggleSettings} />
                    {/*<ListComponent itemList={this.state.itemList} settings1={this.state.settings1} />*/}
                    {/*<ng-include src="'views/parserVisualiserView.html'"></ng-include>*/}
                </div>
                <div className="col s8 ">
                    {/*<ng-include src="'views/debugger/parserDebuggerView.html'"></ng-include>*/}
                </div>
            </div>);
    }
}
