// EditorComponent/src/EditorApp.tsx
import { PluginManagerFactory } from './plugins/PluginManagerFactory';
import { PluginManager } from './plugins/PluginManager';
import { EditorSettingsService } from './service/EditorSettingsService';
import { EditorSettingsStore } from './store/EditorSettingsStore';
import { EditorSettingsComponent } from "./components/EditorSettingsComponent"
import { MonacoEditorWrapper } from "./components/MonacoEditorWrapper"
import { ParsingService, ParserResult, ParserResultStore } from "./service/ParsingService";
import { observable } from "EditorComponent/node_modules/mobx/lib/mobx";
import { ParserLogger } from "./components/ParserLogger";
import { ErrorConsole } from "./components/ErrorConsole";
import { ErrorCheckPlugin } from "./plugins/ErrorCheckPlugin/ErrorCheckPlugin";
import { ParserErrorStore } from "./store/ParserErrorsStore";
import { ToolRenderer } from "./components/ToolRenderer";
import { ToolStore, Tool } from "./store/ToolStore";
import { EditorPluginApi } from "./plugins/EditorPluginApi";
import * as React from "react";
import { observer } from "mobx-react";
import DevTools from 'mobx-react-devtools';
import RaisedButton from 'material-ui/RaisedButton';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import { ERSParserDebuggable } from "ERSParser";
var injectTapEventPlugin = require("react-tap-event-plugin");

@observer
export class EditorApp extends React.Component<any, any> {
    private toolStore = new ToolStore();
    private editorSettingsService = new EditorSettingsService();
    private editorSettingsStore: EditorSettingsStore;
    private pluginManager: PluginManager;
    private parserErrorStore: ParserErrorStore;

    constructor() {
        super();
        injectTapEventPlugin();
        this.editorSettingsStore = new EditorSettingsStore();
        this.init();
    }

    private init() {
        this.editorSettingsStore = this.editorSettingsService.createAndLoadSettingsFromCookies();
        var pluginManagerFactory = new PluginManagerFactory();
        this.pluginManager = pluginManagerFactory.createPluginManager();
        this.parserErrorStore = new ParserErrorStore();
    }

    setMonacoEditor(monacoEditor: monaco.editor.IStandaloneCodeEditor) {
        var editorPluginApi = new EditorPluginApi(
            monacoEditor,
            this.editorSettingsStore,
            this.toolStore,
            this.parserErrorStore);

        this.pluginManager.initPlugins(editorPluginApi);
    }

    render() {
        return (<MuiThemeProvider>
            <ToolRenderer toolStore={this.toolStore}>
                <MonacoEditorWrapper setMonacoEditor={this.setMonacoEditor.bind(this)} />
                <ErrorConsole parserErrorStore={this.parserErrorStore} />
                <EditorSettingsComponent editorSettingsStore={this.editorSettingsStore} />
                <DevTools />
            </ToolRenderer>
        </MuiThemeProvider >);
    }
}
