// EditorComponent/src/EditorApp.tsx
import { MonacoEditorProvider } from './providers/MonacoEditorProvider';
import { ErrorMessageProvider } from './providers/ErrorMessageProvider';
import { PluginManagerFactory } from './plugins/PluginManagerFactory';
import { PluginManager } from './plugins/PluginManager';
import { EditorSettingsService } from './service/EditorSettingsService';
import { EditorSettingsStore } from './store/EditorSettingsStore';
import { EditorSettingsComponent } from "./components/EditorSettingsComponent"
import { MonacoEditorWrapper } from "./components/MonacoEditorWrapper"
import { ParsingService, iresult, ParserResultStore } from "./service/ParsingService";
import * as React from "react";
import { observer } from "mobx-react";
import DevTools from 'mobx-react-devtools';
import { observable } from "EditorComponent/node_modules/mobx/lib/mobx";
import { ParserLogger } from "./components/ParserLogger";
import RaisedButton from 'material-ui/RaisedButton';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import { ParserVisualiser } from "./components/ParserVisualiser";
import { ParserVisualisingService, IParserVisualisingServiceResult } from "./service/ParserVisualisingService"
import { ERSParser } from "ERSParser";
var injectTapEventPlugin = require("react-tap-event-plugin");

@observer
export class EditorApp extends React.Component<any, any> {
    private parserVisualisingServiceResult: IParserVisualisingServiceResult;
    private editorSettingsService = new EditorSettingsService();
    private editorSettingsStore: EditorSettingsStore;
    private monacoEditor: monaco.editor.IStandaloneCodeEditor;
    private parserResultStore: ParserResultStore;
    private pluginManager: PluginManager;
    private errorMessageProvider: ErrorMessageProvider;
    private monacoEditorProvider: MonacoEditorProvider;
    @observable someText = "";

    constructor() {
        super();
        injectTapEventPlugin();
        this.editorSettingsStore = new EditorSettingsStore();
        this.init();
        this.runNewParserClick = this.runNewParserClick.bind(this);
        this.setMonacoEditor = this.setMonacoEditor.bind(this);
    }

    private init() {
        this.editorSettingsStore = this.editorSettingsService.createAndLoadSettingsFromCookies();
        this.parserResultStore = new ParserResultStore();
        var parserVisualisingService = new ParserVisualisingService();
        this.parserVisualisingServiceResult = parserVisualisingService.run(new ERSParser());
        var pluginManagerFactory = new PluginManagerFactory();
        this.pluginManager = pluginManagerFactory.createPluginManager();
        this.errorMessageProvider = new ErrorMessageProvider();
        this.monacoEditorProvider = new MonacoEditorProvider();
    }

    runNewParserClick() {
        var parserService = new ParsingService(this.editorSettingsStore);
        var res = parserService.run(this.monacoEditor.getModel().getValue());
        this.errorMessageProvider.setEditorErrors(res.parserErrors)
        this.parserResultStore.result = res;
    }

    setMonacoEditor(monacoEditor: monaco.editor.IStandaloneCodeEditor) {
        this.monacoEditor = monacoEditor;
        this.pluginManager.initPlugins(monacoEditor);
        this.errorMessageProvider.init(monacoEditor);
    }

    render() {
        return (<MuiThemeProvider>
            <div className="row">
                <div className="col s4">
                    <MonacoEditorWrapper setMonacoEditor={this.setMonacoEditor} monacoEditorProvider={this.monacoEditorProvider} />
                    <RaisedButton onClick={this.runNewParserClick}>Run</RaisedButton>
                    <EditorSettingsComponent editorSettingsStore={this.editorSettingsStore} />
                    <ParserVisualiser parserVisualiser={this.parserVisualisingServiceResult} ediorSettingsStore={this.editorSettingsStore} />
                </div>
                <div className="col s8 ">
                    <ParserLogger parserResultStore={this.parserResultStore} editorSettingsStore={this.editorSettingsStore} />
                </div>
                <DevTools />
            </div>
        </MuiThemeProvider>);
    }
}
