// EditorComponent/src/plugins/AutocompletitionPlugin/AutocompletitionPlugin.ts
import { IEditorPluginApi } from '../EditorPluginApi';
import { IPlugin } from '../IPlugin';
import { MonacoSparqlCompletionItemProvider } from './MonacoSparqlCompletionItemProvider';
import { SparqlSignatureHelpProvider, SparqlLanguageServer } from "SparqlAutocompletion";

export class signatureprovider implements monaco.languages.SignatureHelpProvider {
    signatureHelpTriggerCharacters: string[] = [' '];
    sparqlLanguageServer:SparqlLanguageServer;

    constructor(sparqlLanguageServer: SparqlLanguageServer) {
        this.sparqlLanguageServer = sparqlLanguageServer;
    }

    createHackSignatureHelp(label: string, documentation: string): monaco.languages.SignatureHelp {
        return {
            signatures: [{
                label: "Expected",
                parameters: [{
                    documentation: label,
                    label: "label"
                }]
            }],
            activeParameter: 0,
            activeSignature: 0
        }
    }

    provideSignatureHelp(model: monaco.editor.IReadOnlyModel, position: monaco.Position, token: monaco.CancellationToken): monaco.languages.SignatureHelp {
        var signatureHelp = this.sparqlLanguageServer.sparqlSignatureHelpProvider.provideSignatureHelp(model.getValue(), model.getOffsetAt(position));
        if (signatureHelp == null) {
            return null;
        }
        return this.createHackSignatureHelp(signatureHelp.label, signatureHelp.documentation);
    }
}

export class AutocompletitionPlugin implements IPlugin {
    init(editorPluginApi: IEditorPluginApi): void {
        var sparqlLanguageServer = editorPluginApi.getSparqlLanguageServerService().getLanguageServer();
        monaco.languages.registerCompletionItemProvider('sparql', new MonacoSparqlCompletionItemProvider(sparqlLanguageServer));
        monaco.languages.registerSignatureHelpProvider('sparql', new signatureprovider(sparqlLanguageServer))
    }
}
