// EditorComponent/src/plugins/AutocompletitionPlugin/MonacoSparqlCompletionItemProvider.ts
import { SparqlAutocompleteProvider, OntologyHiearchy, OntologyAssertionsBuilder, OntologyHiearchyBuilder, TurtleGraphWrapper } from "SparqlAutocompletion";
import { ERSParser } from 'ERSParser';

export class MonacoSparqlCompletionItemProvider implements monaco.languages.CompletionItemProvider {
    private sparqlAutocompleteProvider: SparqlAutocompleteProvider;
    
    constructor() {
        var ontologyContent = require('raw-loader!../../resources/dbpedia.ttl');
        this.sparqlAutocompleteProvider = new SparqlAutocompleteProvider(ontologyContent);
    }

    triggerCharacters: string[] = [];

    provideCompletionItems(model: monaco.editor.IReadOnlyModel, position: monaco.Position): monaco.languages.CompletionItem[] {
        return this.sparqlAutocompleteProvider
            .provideAutocompleteItems(model.getValue(), model.getOffsetAt(position))
            .map(x => ({
                label: x.label,
                insertText: x.insertText,
                kind: <any>x.kind
            }));
    }
}
