// EditorComponent/src/plugins/AutocompletitionPlugin/SparqlCompletionItemProvider.ts
import { AutocompleteProviderClass, OntologyHiearchy, OntologyAssertionsBuilder, OntologyHiearchyBuilder, TurtleGraphWrapper } from "AutocompleteProvider";
import ERSParser from 'ERSParser';

export class SparqlCompletionItemProvider implements monaco.languages.CompletionItemProvider {
    private ontologyAssertionsBuilder = new OntologyAssertionsBuilder();
    private ontologyHiarchyBuilder = new OntologyHiearchyBuilder();
    private ontologyHiearchy: OntologyHiearchy;

    triggerCharacters: string[] = [];

    constructor() {
        console.log('Start preparing ontology hiearchy.')
        this.ontologyHiearchy = this.createOntologyHiearchyFromFile();
        console.log('Ontology hiearchy prepared.')
    }

    createDependencyProposals() {
        return [
            {
                label: '"label"',
                kind: monaco.languages.CompletionItemKind.Function,
                documentation: '"DOc"',
                insertText: 'label',
                detail: '"detail"',
                filterText: 'bagr',
                sortText: "1"
            },
            {
                label: '"ns:jead"',
                kind: monaco.languages.CompletionItemKind.Function,
                documentation: "Fast, unopinionated, minimalist web framework",
                insertText: '"express": "*"',
                sortText: "2"
            },
            {
                label: '"ns:dva"',
                kind: monaco.languages.CompletionItemKind.Function,
                documentation: "Fast, unopinionated, minimalist web framework",
                insertText: '"express": "*"',
                sortText: "3"
            },
            {
                label: '"express"',
                kind: monaco.languages.CompletionItemKind.Function,
                documentation: "Fast, unopinionated, minimalist web framework",
                insertText: '"express": "*"',
                sortText: "4"
            },
            {
                label: '"mkdirp"',
                kind: monaco.languages.CompletionItemKind.Function,
                documentation: "Recursively mkdir, like <code>mkdir -p</code>",
                insertText: '"mkdirp": "*"',
                sortText: "5"
            },
            {
                label: '"lodash"',
                kind: monaco.languages.CompletionItemKind.Function,
                insertText: '"lodash": "*"',
                sortText: "6"
            },
        ];
    }

    takeWhileNotWhitespace(text: string) {
        for (var i = text.length - 1; i >= 0; i--) {
            if (text[i] == " ") {
                return text.substr(0, i);
            }
        }

        return "";
    }

    injectCursor(query: string, offset: number) {
        return this.takeWhileNotWhitespace(query.substring(0, offset)) + ' ?cursor ' + query.substring(offset, query.length);
    }

    createOntologyHiearchyFromFile() {
        var fileContent = require('raw-loader!../../resources/dbpedia.ttl');
        var ontologyAssertions = this.ontologyAssertionsBuilder.buildOntologyAsertions(new TurtleGraphWrapper(fileContent));
        var ontologyHiearchy = this.ontologyHiarchyBuilder.createOntologyHiearchy(ontologyAssertions);
        return ontologyHiearchy;
    }

    parseQueryFromFile(queryContent: string) {
        var parser = new ERSParser();
        return parser.parse(queryContent);
    }

    doAutocomplete(queryContent: string) {
        var autocompleteProvider = new AutocompleteProviderClass(this.ontologyHiearchy);
        var query = this.parseQueryFromFile(queryContent);
        var autocompleteResult = autocompleteProvider.doAutocomplete(query);
        console.log(autocompleteResult);
        return autocompleteResult;
    }

    provideCompletionItems(model: monaco.editor.IReadOnlyModel, position: monaco.Position, token: monaco.CancellationToken): monaco.languages.CompletionItem[] {
        var query = this.injectCursor(model.getValue(), model.getOffsetAt(position));
        console.log(query);

        var result = this.doAutocomplete(query);
        var cursorContext = result.getVariableContext('cursor');
        if (cursorContext != null) {
            return cursorContext.possibleAdvices.map(x => ({
                label: x,
                kind: monaco.languages.CompletionItemKind.Function,
                documentation: "None sadfdsafsad fsad sad fdsaf sad fsad fas dfsadfsadfsadfsaddsafs d sadfsdfsd",
                detail: "Documenta asdf asdf adsf asdf asdfasdfadsfas",
                insertText: '<' + x + '>',
                sortText: "2"
            }));
        }
        return [];
    }
}
