// SparqlAutocompletion/src/Serialization/TypeSerializers/OntologyClassSerializer.ts
import { ListClass, IOntologyClass, AtomicClass, UnknownClass, UnionClass, IntersectionClass, RestrictionClass, ComplementClass } from '../../Ontology/OntologyClass';
import { SerializationContext, BaseTypeSerializer } from '../TypeSerializer';
var uuid = require("uuid");

/**
 * Provide serialization of OntologyClass
 */
export class OntologyClassSerializer extends BaseTypeSerializer<IOntologyClass> {
    serialize(serializeObject: IOntologyClass, serializationContext: SerializationContext) {
        var result: any;

        if (serializeObject instanceof AtomicClass) {
            result = {
                type: 0,
                iri: serializeObject.iri
            }
        }
        else if (serializeObject instanceof RestrictionClass) {
            result = {
                type: 1,
                onProperty: serializeObject.onProperty
            };
        }
        else if (serializeObject instanceof ListClass) {
            var isUnionClass = serializeObject instanceof UnionClass;
            var serializedList: any[] = serializeObject.classes.map(x => this.serialize(x, serializationContext));
            result = {
                type: isUnionClass ? 2 : 3,
                classes: serializedList
            }
        }
        else if (serializeObject instanceof UnknownClass) {
            result = {
                type: 4,
                nodeValue: serializeObject.nodeValue
            }
        }
        else if (serializeObject instanceof ComplementClass) {
            result = {
                type: 5,
                cclass: this.serialize(serializeObject.value, serializationContext)
            }
        }
        else {
            throw new Error("Unknown class type to serialization.")
        }

        if (serializeObject.label.containsAnyLabel()) {
            var ontologyLabelSerializer = serializationContext.getTypeSerializer("OntologyLabel");
            result.label = ontologyLabelSerializer.serializeCommon(serializeObject.label, serializationContext);
        }

        return result;
    }

    deserializeClass(payload: any, serializationContext: SerializationContext): IOntologyClass {
        switch (payload.type) {
            case 0:
                return new AtomicClass(payload.iri);
            case 1:
                return new RestrictionClass(payload.onProperty);
            case 2:
            case 3:
                var result = payload.type == 2 ? new UnionClass() : new IntersectionClass();
                payload.classes.forEach((serializedClass: any) => result.addClass(this.deserialize(serializedClass, serializationContext)));
                return result;
            case 4:
                return new UnknownClass(payload.nodeValue);
            case 5:                
                return new ComplementClass(this.deserialize(payload.cclass, serializationContext));
            default:
                throw Error("Unkown ontology class type.");
        }
    }

    deserialize(payload: any, serializationContext: SerializationContext): IOntologyClass {
        var deserializedClass = this.deserializeClass(payload, serializationContext);
        if (payload.label !== undefined) {
            var ontologyLabelSerializer = serializationContext.getTypeSerializer("OntologyLabel");
            deserializedClass.label = ontologyLabelSerializer.deserializeCommon(payload.label, serializationContext);
        }
        return deserializedClass;
    }
}
