// SparqlAutocompletion/src/Ontology/OntologyClassTextFormatter.ts
import { AtomicClass, IntersectionClass, UnionClass, UnknownClass, RestrictionClass, ListClass, ComplementClass } from "./OntologyClass";

export interface IOntologyClassTextFormatter {
    formatUnknownClass(unknownClass: UnknownClass): string;
    formatAtomicClass(atomicClass: AtomicClass): string;
    formatRestrictionClass(restrictionClass: RestrictionClass): string;
    formatIntersectionClass(intersectionClass: IntersectionClass): string;
    formatUnionClass(unionClass: UnionClass): string;
    formatComplementClass(complementClass: ComplementClass): string;
}

export class SimpleOntologyClassTextFormatter implements IOntologyClassTextFormatter {

    private formatListClasses(listClass: ListClass) {
        return listClass.classes.map(x => x.getText(this)).join(', ');
    }

    formatUnknownClass(unknownClass: UnknownClass): string {
        return 'Unknown(' + unknownClass.nodeValue + ')';
    }

    formatAtomicClass(atomicClass: AtomicClass): string {
        return 'Atomic(' + atomicClass.iri + ')';
    }

    formatRestrictionClass(restrictionClass: RestrictionClass): string {
        return 'RestrictionOn(' + restrictionClass.onProperty + ')';
    }

    formatIntersectionClass(intersectionClass: IntersectionClass): string {
        return 'IntersectionOf(' + this.formatListClasses(intersectionClass) + ')';

    }

    formatUnionClass(unionClass: UnionClass): string {
        return 'UnionOf(' + this.formatListClasses(unionClass) + ')';
    }

    formatComplementClass(complementClass: ComplementClass): string {
        return 'ComplementOf(' + complementClass.value.getText(this) + ')';
    }
}

export class CleanOntologyClassTextFormatter implements IOntologyClassTextFormatter {

    private formatListClasses(listClass: ListClass) {
        return listClass.classes.map(x => x.getText(this));
    }

    formatUnknownClass(unknownClass: UnknownClass): string {
        return 'Unknown class (' + unknownClass.nodeValue + ')';
    }

    formatAtomicClass(atomicClass: AtomicClass): string {
        return atomicClass.iri;
    }

    formatRestrictionClass(restrictionClass: RestrictionClass): string {
        return 'Restriction on ' + restrictionClass.onProperty;
    }

    formatIntersectionClass(intersectionClass: IntersectionClass): string {
        return this.formatListClasses(intersectionClass).join(' and ');
    }

    formatUnionClass(unionClass: UnionClass): string {
        return this.formatListClasses(unionClass).join(' or ');
    }

    formatComplementClass(complementClass: ComplementClass): string {
        return 'Complement of (' + complementClass.value.getText(this) + ')';
    }
}
