// EditorComponent/webpack.config.js
var bitBarWebpackProgressPlugin = require("bitbar-webpack-progress-plugin");
var CopyWebpackPlugin = require("copy-webpack-plugin");
var webpack = require("webpack");

module.exports = {
    entry: {
        release: "./src/indexRelease.ts",
        debug: [
            "webpack-dev-server/client?http://localhost:8050",
            "webpack/hot/only-dev-server",
            "./src/indexDebug.ts"]
    },
    output: {
        path: __dirname,
        filename: "dist/[name]bundle.js",
        devtoolModuleFilenameTemplate: "../[resource-path]",
        libraryTarget: "amd"
    },
    devServer: {
        port: 8050,
        hot: true
    },
    resolve: {
        extensions: [".js", ".ts", ".tsx", ".css"]
    },
    devtool: "source-map",
    module: {
        rules: [
            {
                test: /\.tsx?$/,
                loader: "ts-loader"
            },
            {
                enforce: "pre",
                test: /\.js$/,
                loader: "source-map-loader"
            },
            {
                test: /\.css$/,
                use: [
                    { loader: "style-loader" },
                    { loader: "css-loader" },
                ],
            }
        ]
    },
    plugins: [
        new webpack.HotModuleReplacementPlugin(),
        new webpack.NamedModulesPlugin(),
        new bitBarWebpackProgressPlugin(),
        new CopyWebpackPlugin([{ from: "../../ontologies", to: "./ontologies" }])
    ]
};
