// EditorComponent/src/plugins/CursorInjectionPlugin/CursorInjectionPlugin.ts
import { IPlugin } from '../IPlugin';
import { injectCursor, injectUnknown } from "SparqlAutocompletion"

export class CursorInjectionPlugin implements IPlugin {
    createAction(id: string,
        keybindings: number[],
        run: (editor: monaco.editor.ICommonCodeEditor) => void) {
        return {
            id: id,
            label: id,
            keybindings: keybindings,
            contextMenuGroupId: 'navigation',
            contextMenuOrder: 1.5,
            run: run
        }
    }

    init(monacoEditor: monaco.editor.IStandaloneCodeEditor): void {
        monacoEditor.addAction(this.createAction(
            'cursorInjectionPlugin',
            [monaco.KeyMod.Alt | monaco.KeyCode.KEY_L],
            function (ed) {
                var query = injectCursor(ed.getModel().getValue(), ed.getModel().getOffsetAt(ed.getPosition()));
                console.log(query);
            }
        ));
        
        monacoEditor.addAction(this.createAction(
            'unknownInjectionPlugin',
            [monaco.KeyMod.Alt | monaco.KeyCode.KEY_K],
            function (ed) {
                var query = injectUnknown(ed.getModel().getValue(), ed.getModel().getOffsetAt(ed.getPosition()));
                console.log(query);
            }
        ));

    }
}
