// EditorComponent/src/react/components/ParserLoggerStateStackView.tsx
import { EditorSettingsStore } from '../store/EditorSettingsStore';
import * as React from "react";
import { observer } from "mobx-react";
import { JsonFormatter } from "./utils/JsonFormatter";

export interface ParserLoggerStateStackViewProps {
    stateStack: any[];
}

export class ParserLoggerStateStackView extends React.Component<ParserLoggerStateStackViewProps, any> {
    renderRule(stackItem: any) {
        return <div>
            <div className="orange lighten-4">{stackItem.Id}</div>
            <div className="black-text">
                <ul className="collection">
                    <li className="collection-item">Rules:
                        <ul className="collection">
                            {stackItem.Rules.map((rule: any, index: number) =>
                                <li key={index} className="collection-item">
                                    {rule}
                                </li>)
                            }
                        </ul>
                    </li>
                    <li className="collection-item">Transitions:
                        <ul className="collection">
                            {
                                Object.keys(stackItem.Transitions).map((key: string, index: number) =>
                                    <li className="collection-item" key={index}>
                                        {key} - &gt; {stackItem.Transitions[key]}
                                    </li>
                                )
                            }
                        </ul>
                    </li>
                </ul>
            </div>
        </div>
    }

    renderSymbol(stackItem: any) {
        return <div>{stackItem}</div>
    }

    render() {
        return <div> State stack:
            <ul className="collection">
                {this.props.stateStack.map((stackItem: any, index: number) =>
                    <li key={index} className="collection-item">
                        {
                            index % 2 == 0 ? this.renderRule(stackItem) : this.renderSymbol(stackItem)
                        }
                    </li>)
                }

            </ul>
        </div>
    }
}
