// EditorComponent/src/react/components/ParserVisualiser.tsx
import * as React from "react";
import { observer } from "mobx-react";
import { ParserVisualisingService, IParserVisualisingServiceResult } from '../service/ParserVisualisingService';
import { Card, CardHeader, CardText, List, ListItem, Paper } from "material-ui"
import { EditorSettingsStore } from "../store/EditorSettingsStore";

export interface ParserVisualiserProps {
    parserVisualiser: IParserVisualisingServiceResult;
    ediorSettingsStore: EditorSettingsStore;
}

@observer
export class ParserVisualiser extends React.Component<ParserVisualiserProps, any> {
    private parserVisualisingService = new ParserVisualisingService();

    references(itemSet: any) {
        return this.parserVisualisingService.getItemSetReferences(this.props.parserVisualiser.itemSets, itemSet.Id)
            .map((referenceId: any, index: number) => (
                <ListItem key={index}>
                    <a href={'#itemSet' + referenceId} target="_self">{referenceId}</a>
                </ListItem>
            ));

    }

    rules(itemSet: any) {
        return itemSet.Rules.map((rule: any, index: number) => <ListItem key={index}>{rule}</ListItem>)
    }

    transitions(itemSet: any) {
        return Object.keys(itemSet.Transitions).map((transitionSymbol: any, index: number) => (
            <ListItem key={index}>
                <a href="#itemSet{{transitionState}}" target="_self">{transitionSymbol} - {itemSet.Transitions[transitionSymbol]}</a>
            </ListItem>
        ));
    }

    renderAction(action: any) {
        switch (action.actionType) {
            case 'goto':
                return <div>{action.symbol} Goto {action.state}</div>;
            case 'shift':
                return <div>{action.symbol} Shift {action.state}</div>;
            case 'reduce':
                return <div>{action.symbol} Reduce {action.production}</div>;
            default:
                throw new Error('Unknonw action type.');
        }
    }
    table(itemSet: any) {
        return this.props.parserVisualiser.table[itemSet.Id].map((action: any, index: number) => (
            <ListItem key={index}>
                {this.renderAction(action)}
            </ListItem>
        ));
    }

    renderItemSet(itemSet: any, index: number) {
        return (
            <Card key={index}>
                <CardHeader title={itemSet.Id} actAsExpander={true} />
                <CardText expandable={true}>
                    <List>
                        <ListItem>
                            Rules:
                            <List>
                                {this.rules(itemSet)}
                            </List>
                        </ListItem>
                        <ListItem>
                            References:
                            <List>
                                {this.references(itemSet)}
                            </List>
                        </ListItem>
                        <ListItem>
                            Transitions:
                            <List>
                                {this.transitions(itemSet)}
                            </List>
                        </ListItem>
                        <ListItem>
                            Table:
                            <List>
                                {this.table(itemSet)}
                            </List>
                        </ListItem>
                    </List>
                </CardText>
            </Card>
        );
    }

    render() {
        if (!this.props.ediorSettingsStore.getSettings('showParserStates')) {
            return null;
        }
        return (
            <Card>
                <CardHeader title="Parser states" />
                <CardText>
                    {
                        this.props.parserVisualiser.itemSets.map((itemSet: any, index: number) => this.renderItemSet(itemSet, index))
                    }
                </CardText>
            </Card>
        );
    }
}
