// EditorComponent/src/plugins/SyntaxHighlightPlugin/SyntaxHighlightPlugin.ts
import { ErrorMessageProvider } from '../../providers/ErrorMessageProvider';
import { IPlugin, IEditorPluginApi } from '../IPlugin';
import { ERSParser } from "ERSParser"

class SparqlMonarchDefinition implements monaco.languages.IMonarchLanguage {

	keywords = ["base", "prefix", "select", "distinct", "reduced", "as", "construct", "where", "describe", "ask",
		"from", "named", "group", "by", "having", "order", "asc", "desc", "limit", "offset", "values", "load", "silent",
		"into", "clear", "drop", "create", "add", "to", "move", "copy", "insert", "delete", "delete", "with", "delete", "insert",
		"using", "default", "graph", "all", "optional", "service", "bind", "undef", "minus", "union", "filter", "a", "in", "not",
		"bound", "bnode", "concat", "coalesce", "if", "regex", "substr", "replace", "exists", "count", "group_concat",
		"separator", "true", "false",
	];
	keywordsUpperCase = ["BASE", "PREFIX", "SELECT", "DISTINCT", "REDUCED", "AS", "CONSTRUCT", "WHERE", "DESCRIBE", "ASK",
		"FROM", "NAMED", "GROUP", "BY", "HAVING", "ORDER", "ASC", "DESC", "LIMIT", "OFFSET", "VALUES", "LOAD", "SILENT",
		"INTO", "CLEAR", "DROP", "CREATE", "ADD", "TO", "MOVE", "COPY", "INSERT", "DELETE", "DELETE", "WITH", "DELETE", "INSERT",
		"USING", "DEFAULT", "GRAPH", "ALL", "OPTIONAL", "SERVICE", "BIND", "UNDEF", "MINUS", "UNION", "FILTER", "A", "IN", "NOT",
		"BOUND", "BNODE", "CONCAT", "COALESCE", "IF", "REGEX", "SUBSTR", "REPLACE", "EXISTS", "COUNT", "GROUP_CONCAT",
		"SEPARATOR", "TRUE", "FALSE",
	];

	pn_chars_base = /[A-Z]|[a-z]|[\u00C0-\u00D6]|[\u00D8-\u00F6]|[\u00F8-\u02FF]|[\u0370-\u037D]|[\u037F-\u1FFF]|[\u200C-\u200D]|[\u2070-\u218F]|[\u2C00-\u2FEF]|[\u3001-\uD7FF]|[\uF900-\uFDCF]|[\uFDF0-\uFFFD]|[\uD800-\uDB7F][\uDC00-\uDFFF]/;

	pn_chars_u = /(?:(@pn_chars_base)|"_")/;;

	varname = /((@pn_chars_u)|[0-9])((@pn_chars_u)|[0-9]|\u00B7|[\u0300-\u036F]|[\u203F-\u2040])*/;

	tokenizer: { [name: string]: monaco.languages.IMonarchLanguageRule[]; } = {
		root: [
			[/[\?\$](@varname)/, "variable"],
			[/[a-zA-z0-9]+/, {
				cases: {
					'@keywords': "keyword",
					'@keywordsUpperCase': "keyword",
					"@default": "identifier"
				}
			}],
			{ include: '@comment' },
		],
		comment: [
			[/#[^\n\r]*/, 'comment']
		]
	};


	ignoreCase: true;
	defaultToken: "INVALID";
	tokenPostfix: "";
}

export class SyntaxHighlightPlugin implements IPlugin {
	init(monacoEditor: monaco.editor.IStandaloneCodeEditor, editorPluginApi: IEditorPluginApi): void {
		var sparqlMonarchDefinition = new SparqlMonarchDefinition();
		monaco.languages.setMonarchTokensProvider('sparql', sparqlMonarchDefinition);

		monaco.languages.setLanguageConfiguration("sparql", {
			autoClosingPairs: [{ open: "{", close: "}" },
			{ open: "(", close: ")" }],
			brackets: [["{", "}"],
			["(", ")"]],

		});
		monaco.editor.defineTheme('sparqlTheme', {
			base: 'vs',
			inherit: true,
			rules: [
				{ token: 'variable', foreground: 'BBA600' },
				{ token: 'comment', foreground: 'BBA6FF' },
			]
		});
		monacoEditor.updateOptions({ theme: 'sparqlTheme' });
	}
}
