// EditorComponent/src/providers/ErrorMessageProvider.ts
import { ParserError } from "ERSParser";

export class ErrorMessageProvider {
    private monacoEditorModel: monaco.editor.IModel;

    init(monacoEditor: monaco.editor.IStandaloneCodeEditor): void {
        this.monacoEditorModel = monacoEditor.getModel();
    }

    findErrorNewStartOffset(editorContent: string, offsetOfStart: number, offsetOfEnd: number) {
        for (var i = offsetOfStart + 1; i < offsetOfEnd - 1; i++) {
            var currentChar = editorContent[i];
            if (currentChar != " " && currentChar != "\r" && currentChar != "\n") {
                return i;
            }
        }

        return offsetOfStart;
    }

    skipSpacesOnErrorBeginning(error: any): monaco.Position {
        var editorModel = this.monacoEditorModel;
        var editorContent = editorModel.getValue();

        var offsetOfStart = editorModel.getOffsetAt(new monaco.Position(
            error.yylocStart.last_line,
            error.yylocStart.last_column));

        var offsetOfEnd = editorModel.getOffsetAt(new monaco.Position(
            error.yylocEnd.last_line,
            error.yylocEnd.last_column,
        ));

        var newStartOffset = this.findErrorNewStartOffset(editorContent, offsetOfStart, offsetOfEnd);
        var newStartPosition = editorModel.getPositionAt(newStartOffset);
        return newStartPosition;
    }

    setEditorErrors(errors: ParserError[]) {
        var markers = errors.map(error => {
            var errorStartPosition = this.skipSpacesOnErrorBeginning(error);
            return {
                severity: monaco.Severity.Error,
                code: '',
                source: '',
                startLineNumber: errorStartPosition.lineNumber,
                startColumn: errorStartPosition.column,
                endLineNumber: error.yylocEnd.last_line,
                endColumn: error.yylocEnd.last_column,
                message: error.type,
            }
        });
        monaco.editor.setModelMarkers(this.monacoEditorModel, 'sparql', markers);
    }
}
