// EditorComponent/src/plugins/Tools/PropertyAutocompleteTool/components/NodeComponents/DataModel.ts
import { PrefixProcessor, OntologyAssertions, OntologyConcept, CleanOntologyClassTextFormatter } from "SparqlAutocompletion";

export interface AutocompleteClass {
    name: string;
    properties: string[];
}

export interface AutocompleteConcept {
    classes: AutocompleteClass[];
}

export interface ClassNodeInput {
    classes: AutocompleteClass[];
}

export interface ConceptNodeInput {
    concepts: AutocompleteConcept[];
}

export interface PropertyNodeInput {
    ontologyProperties: string[];
}

export function createConceptNodeInput(ontologyAssertions: OntologyAssertions, prefixProcessor: PrefixProcessor, ontologyConcepts: OntologyConcept[]): ConceptNodeInput {
    var ontologyClassTextFormatter = new CleanOntologyClassTextFormatter();
    var ontologyAssertions = ontologyAssertions;

    var autocompleteConcepts = ontologyConcepts.map(ontologyConcept => {
        var autocompleteClasses = ontologyConcept.ontologyClasses.map(ontologyClass => {
            var properties = ontologyAssertions.findPropertiesByDomain(ontologyClass).map(x => prefixProcessor.processPrefixesInIri(x.name))
            return {
                name: ontologyClass.getText(ontologyClassTextFormatter),
                properties: properties
            }
        });

        return {
            classes: autocompleteClasses
        };
    });


    return {
        concepts: autocompleteConcepts
    };
}
