// EditorComponent/src/plugins/Tools/PropertyAutocompleteTool/components/NodeComponents/NodeFactory.tsx
import * as React from "react";
import { PropertyAutocompleteOptions } from "../../PropertyAutocompleteOptions";

export interface INodeFactoryContext {
    renderListUnder(data: any): JSX.Element;
    propertyAutocompleteOptions: PropertyAutocompleteOptions;
    insertPropertyIntoEditor(propertyName: string): void
}

export interface INodeFactoryCommon {
    dataSelectorCommon(data: any, nodeFactoryContext: INodeFactoryContext): any;
    getListElementCommon(data: any, nodeFactoryContext: INodeFactoryContext): JSX.Element;
}

export interface INodeFactory<TFrom, TTo> extends INodeFactoryCommon {
    dataSelector(data: TFrom, nodeFactoryContext: INodeFactoryContext): TTo;
    getListElement(data: TFrom, nodeFactoryContext: INodeFactoryContext): JSX.Element;
}

export abstract class BaseNodeFactory<TFrom, TTo> implements INodeFactory<TFrom, TTo> {
    public dataSelectorCommon(data: any, nodeFactoryContext: INodeFactoryContext): any {
        return this.dataSelector(data, nodeFactoryContext);
    }

    public getListElementCommon(data: any, nodeFactoryContext: INodeFactoryContext): JSX.Element {
        return this.getListElement(data, nodeFactoryContext);
    }

    abstract dataSelector(data: TFrom, nodeFactoryContext: INodeFactoryContext): TTo;

    abstract getListElement(data: TFrom, nodeFactoryContext: INodeFactoryContext): JSX.Element;
}
