// EditorComponent/src/plugins/Tools/PropertyAutocompleteTool/components/NodeComponents/PropertyNode.tsx
import * as React from "react";
import { observer } from "mobx-react";
import { List, ListItem, Divider, RaisedButton, Paper, Card, CardHeader, CardText } from "material-ui"
import { observable } from "mobx";
import { SparqlLanguageServer, OntologyAutocompleteSubProvider, OntologyConcept, OntologyProperty, OntologyHiearchy } from "SparqlAutocompletion";
import { CleanOntologyClassTextFormatter } from "SparqlAutocompletion";
import { IOntologyClass, } from "SparqlAutocompletion";
import { INodeFactoryContext, BaseNodeFactory } from "./NodeFactory";
import { ExpandableList, ExpandableListItem } from "../ExpandableList";
import { PropertyNodeInput } from "./DataModel";

export interface PropertyNodeProps {
    ontologyProperties: OntologyProperty[];
    searchText: string;
}

export class PropertyNodeComponent extends React.Component<PropertyNodeProps, any> {

    render() {
        var searchText = this.props.searchText;
        var properties = this.props.ontologyProperties;

        if (properties.length == 0) {
            return <div>No properties</div>
        }
        return (
            <ExpandableList>
                {properties.map((property, propertyIndex) => {
                    return (
                        <ExpandableListItem key={propertyIndex}>
                            {property.name}
                        </ExpandableListItem>
                    );
                })
                }
            </ExpandableList>
        );
    }
}

export class PropertyNodeFactory extends BaseNodeFactory<PropertyNodeInput, any> {

    dataSelector(data: PropertyNodeInput, nodeFactoryContext: INodeFactoryContext) {
        throw new Error('Final node factory do not select data!');
    }

    getListElement(data: PropertyNodeInput, nodeFactoryContext: INodeFactoryContext): JSX.Element {
        return <PropertyNodeComponent
            searchText={nodeFactoryContext.propertyAutocompleteOptions.searchText}
            ontologyProperties={data.ontologyProperties} />
    }
}
