// EditorComponent/src/plugins/SparqlLanguagePlugin/Providers/SignatureHelpProvider.ts
import { SparqlLanguageServer } from "SparqlAutocompletion";

export class SignatureHelpProvider implements monaco.languages.SignatureHelpProvider {
    signatureHelpTriggerCharacters: string[] = [' '];
    sparqlLanguageServer: SparqlLanguageServer;

    constructor(sparqlLanguageServer: SparqlLanguageServer) {
        this.sparqlLanguageServer = sparqlLanguageServer;
    }

    createHackSignatureHelp(label: string, documentation: string): monaco.languages.SignatureHelp {
        return {
            signatures: [{
                label: "Expected",
                parameters: [{
                    documentation: label,
                    label: "label"
                }]
            }],
            activeParameter: 0,
            activeSignature: 0
        }
    }

    provideSignatureHelp(model: monaco.editor.IReadOnlyModel, position: monaco.Position, token: monaco.CancellationToken): monaco.languages.SignatureHelp {
        var signatureHelpProvider = this.sparqlLanguageServer.sparqlSignatureHelpProvider;
        var text = model.getValue();
        var offset = model.getOffsetAt(position);
        var signatureHelp = signatureHelpProvider.provideSignatureHelp(text, offset);
        if (signatureHelp == null) {
            return null;
        }
        return this.createHackSignatureHelp(signatureHelp.label, signatureHelp.documentation);
    }
}
