// EditorComponent/src/plugins/Tools/PropertyAutocompleteTool/components/ExpandableList.tsx
import * as React from "react";
import { observer } from "mobx-react";
import { observable } from "mobx";

export interface ExpandableListItemProps {
    className?: string;
    onClick?: React.EventHandler<React.MouseEvent<HTMLLIElement>>;
}

export interface ExpandableListProps {
    className?: string;
    showMoreClassName?: string;
    previewItemCount?: number;
}

enum ExpandableListState { Open, Close }

export class ExpandableListItem extends React.Component<ExpandableListItemProps, any> {
    render() {
        return (
            <li className={this.props.className} onClick={this.props.onClick}>{this.props.children}</li>
        );
    }
}

@observer
export class ExpandableList extends React.Component<ExpandableListProps, any> {
    @observable listState: ExpandableListState = ExpandableListState.Close;

    private get previewItemCount() {
        return this.props.previewItemCount || 10;
    }

    toggleState() {
        this.listState = this.listState == ExpandableListState.Open ? ExpandableListState.Close : ExpandableListState.Open;
    }

    render() {
        var children = this.props.children as ExpandableListItem[];
        var isClose = this.listState == ExpandableListState.Close;
        if (isClose) {
            children = children.slice(0, this.previewItemCount);
        }

        return (
            <ul className={this.props.className}>
                {children}
                {children.length < this.previewItemCount ? null :
                    <ExpandableListItem onClick={this.toggleState.bind(this)} className={this.props.showMoreClassName}>
                        Show {isClose ? "more" : "less"} ...
                    </ExpandableListItem>
                }
            </ul>
        );
    }
}
