// SparqlAutocompletion/src/Autocompletition/Inference/QueryAdvice.ts
import { PrefixProcessor } from "../../Autocompletition/Query/PrefixProcessor";
import { IOntologyClass } from "../../Ontology/OntologyClass";
import { OntologyConcept } from "../../Ontology/OntologyConcept";
import { OntologyAssertions, OntologyProperty } from '../../Ontology/OntologyAssertions';
import { StringDictionary } from "../../Collections/Dictionary";

import * as _ from 'lodash';

export interface IQueryAdvice {
    getStringQueryAdvices(): string[]
}

export class RdfTypeQueryAdvice implements IQueryAdvice {
    constructor(private possibleRdfTypes: string[]) {

    }

    getStringQueryAdvices(): string[] {
        return this.possibleRdfTypes;
    }

    getData() {
        return this.possibleRdfTypes;
    }
}

export class PropertyQueryAdvice implements IQueryAdvice {

    public processor: PrefixProcessor;

    private ontologyAssertions: OntologyAssertions;

    constructor(
        public orderConcepts: OntologyConcept[],
        ontologyAssertions: OntologyAssertions,
        prefixes: StringDictionary<string>) {

        this.ontologyAssertions = ontologyAssertions;
        this.processor = new PrefixProcessor(prefixes);
    }

    getAllPoperties() {
        var result = _(this.orderConcepts)
            .map(x => this.getPropertiesOfConcept(x))
            .flatten<OntologyProperty>()
            .value();
        return result;
    }

    getPropertiesOfClass(ontologyClass: IOntologyClass) {
        return this.ontologyAssertions.findPropertiesByDomain(ontologyClass);
    }

    getPropertiesOfConcept(ontologyConcept: OntologyConcept) {
        var result = _(ontologyConcept.ontologyClasses)
            .map(x => this.getPropertiesOfClass(x))
            .flatten<OntologyProperty>()
            .value();

        return result;
    }

    getPropertiesByChunk() {
        return this.orderConcepts.map(x => ({ concept: x, properties: this.getPropertiesOfConcept(x) }))
    }

    getStringQueryAdvices(): string[] {
        var result = _(this.getPropertiesByChunk())
            .map(x => x.properties.map(y => this.processor.processPrefixesInIri(y.name)))
            .flatten<string>()
            .value();

        return result;
    }

    getData(): { concept: OntologyConcept, properties: OntologyProperty[] }[] {
        return this.getPropertiesByChunk();
    }
}
