// EditorComponent/src/service/SparqlProvidersRegistration/Providers/CompletionItemProvider.ts
import { SparqlAutocompleteProvider, SparqlLanguageServer } from "SparqlAutocompletion";
import { SparqlLanguageServerProvider } from "EditorComponent/src/service/SparqlProvidersRegistration/SparqlLanguageServerProvider";

export class CompletionItemProvider implements monaco.languages.CompletionItemProvider {
    private sparqlLanguageServerProvider: SparqlLanguageServerProvider;

    constructor(sparqlLanguageServerProvider: SparqlLanguageServerProvider) {
        this.sparqlLanguageServerProvider = sparqlLanguageServerProvider;
    }

    triggerCharacters: string[] = [];

    provideCompletionItems(model: monaco.editor.IReadOnlyModel, position: monaco.Position): monaco.languages.CompletionItem[] {
        return this.sparqlLanguageServerProvider
            .getSparqlLanguageServer()
            .provideAutocompleteItems(model.getValue(), model.getOffsetAt(position))
            .map(x => ({
                label: x.label,
                insertText: x.insertText,
                kind: <any>x.kind,
                filterText: x.filterText
            }));
    }
}
