// EditorComponent/src/service/SparqlProvidersRegistration/Providers/HoverProvider.ts
import { SparqlLanguageServer } from "SparqlAutocompletion";
import { SparqlLanguageServerProvider } from "EditorComponent/src/service/SparqlProvidersRegistration/SparqlLanguageServerProvider";

export class HoverProvider implements monaco.languages.HoverProvider {
    constructor(private sparqlLanguageServerProvider: SparqlLanguageServerProvider) {
    }

    provideHover(model: monaco.editor.IReadOnlyModel, position: monaco.Position, token: monaco.CancellationToken): monaco.languages.Hover {
        var query = model.getValue();

        var res = this.sparqlLanguageServerProvider.getSparqlLanguageServer().sparqlHoverProvider.provideHover(query, model.getOffsetAt(position));
        if (res == null) {
            return null;
        }
        var result: monaco.languages.Hover = {
            contents: res.contents,
            range: {
                startLineNumber: res.range.startLineNumber,
                startColumn: res.range.startColumn,
                endLineNumber: res.range.endLineNumber,
                endColumn: res.range.endColumn
            }
        };

        return result;
    }
}
