// EditorComponent/src/service/SparqlProvidersRegistration/Providers/SignatureHelpProvider.ts
import { SparqlLanguageServer } from "SparqlAutocompletion";
import { SparqlLanguageServerProvider } from "EditorComponent/src/service/SparqlProvidersRegistration/SparqlLanguageServerProvider";

export class SignatureHelpProvider implements monaco.languages.SignatureHelpProvider {
    signatureHelpTriggerCharacters: string[] = [" "];
    sparqlLanguageServerProvider: SparqlLanguageServerProvider;

    constructor(sparqlLanguageServer: SparqlLanguageServerProvider) {
        this.sparqlLanguageServerProvider = sparqlLanguageServer;
    }

    createHackSignatureHelp(label: string, documentation: string): monaco.languages.SignatureHelp {
        return {
            signatures: [{
                label: "Expected types:",
                parameters: [{
                    documentation: label,
                    label: "label"
                }]
            }],
            activeParameter: 0,
            activeSignature: 0
        };
    }

    provideSignatureHelp(model: monaco.editor.IReadOnlyModel, position: monaco.Position, token: monaco.CancellationToken): monaco.languages.SignatureHelp {
        var signatureHelpProvider = this.sparqlLanguageServerProvider.getSparqlLanguageServer().sparqlSignatureHelpProvider;
        var text = model.getValue();
        var offset = model.getOffsetAt(position);
        var signatureHelp = signatureHelpProvider.provideSignatureHelp(text, offset);
        if (signatureHelp == null) {
            return null;
        }
        return this.createHackSignatureHelp(signatureHelp.label, signatureHelp.documentation);
    }
}
