// SparqlAutocompletion/src/Autocompletition/SparqlLanguageServer.ts
import { Block } from './Query/Query';
import { injectCursor } from './Utils/QueryTerminalInjection';
import { QueryClassInferencer } from './Inference/QueryClassInferencer';
import { QueryAdviceInferencer, PropertyQueryAdvice } from './Inference/QueryAdviceInferencer';
import { OntologyAssertions, OntologyProperty } from '../Ontology/OntologyAssertions';
import { TurtleGraphWrapper } from '../GraphTools/TurtleGraphWrapper';
import { OntologyHierarchy, OntologyConcept } from '../Ontology/OntologyHierarchy';
import { OntologyHierarchyBuilder } from '../Ontology/Algorithms/OntologyHierarchyBuilder';
import { OntologyAssertionsBuilder } from '../Ontology/Algorithms/OntologyAssertionBuilder';
import { SparqlAutocompleteItem, SparqlAutocompleteProvider } from './SparqlAutocompleteProvider';
import { SparqlSignatureHelpProvider } from './SparqlSignatureHelpProvider';
import { SparqlHoverProvider } from 'SparqlAutocompletion/src/Autocompletition/SparqlHoverProvider';
import { Node } from '../Autocompletition/Query/Query'
import { ISparqlLanguageServerSettings, SparqlLanguageServerSettings } from '../Autocompletition/SparqlLanguageServerSettings';
import { EmptyLogger, LoggerFactory } from '../Logging/Logger';
import * as _ from 'lodash';
import { ERSParser } from 'ERSParser';
import { ConsoleLogger } from 'SparqlAutocompletion/src';

export class SparqlLanguageServer {
    private queryAdviceInferencer: QueryAdviceInferencer;

    public _ontologyHierarchy: OntologyHierarchy;
    public _sparqlAutocompleteProvider: SparqlAutocompleteProvider;
    public _sparqlSignatureHelpProvider: SparqlSignatureHelpProvider;
    public _sparqlHoverProvider: SparqlHoverProvider;

    constructor(private ontologyAssertions: OntologyAssertions, sparqlLanguageServerSettings: ISparqlLanguageServerSettings) {
        var ontologyHiarchyBuilder = new OntologyHierarchyBuilder();
        var loggerFactory = new LoggerFactory();
        var coealescedSparqlLanguageServerSettings = new SparqlLanguageServerSettings(sparqlLanguageServerSettings);
        var logger = loggerFactory.createLogger(coealescedSparqlLanguageServerSettings.getIsLoggingEnabled());
        this._ontologyHierarchy = ontologyHiarchyBuilder.createOntologyHierarchy(ontologyAssertions);
        this._sparqlAutocompleteProvider = new SparqlAutocompleteProvider(logger, this._ontologyHierarchy, coealescedSparqlLanguageServerSettings);
        this._sparqlSignatureHelpProvider = new SparqlSignatureHelpProvider(logger, this._ontologyHierarchy);
        this._sparqlHoverProvider = new SparqlHoverProvider(this._ontologyHierarchy, coealescedSparqlLanguageServerSettings.getLanguageTag());
        this.queryAdviceInferencer = new QueryAdviceInferencer(logger, this.ontologyHierarchy, coealescedSparqlLanguageServerSettings.getPropertySelectorType());
    }

    public get ontologyHierarchy() {
        return this._ontologyHierarchy;
    }

    public get sparqlSignatureHelpProvider() {
        return this._sparqlSignatureHelpProvider;
    }

    public get sparqlHoverProvider() {
        return this._sparqlHoverProvider;
    }

    provideAutocompleteItems(query: string, cursorOffset: number): SparqlAutocompleteItem[] {
        return this._sparqlAutocompleteProvider.provideAutocompleteItems(query, cursorOffset);
    }

    provideAutocompleteAdvice(query: string, cursorOffset: number): { concept: OntologyConcept, properties: OntologyProperty[] }[] {
        var propertiesAdvices = this.queryAdviceInferencer.providePropertyAutocompletion(query, cursorOffset);
        var result = propertiesAdvices != null && propertiesAdvices instanceof PropertyQueryAdvice ?
            (propertiesAdvices as PropertyQueryAdvice).getData() : null;
        return result;
    }
}
